File: DataStructures/DoubleLinkedList/cpp/DoubleLinkedList.cpp
#include <stdlib.h>
#include <iostream>
#include <iomanip>
#include <string>
#include <algorithm>
#include <iterator>
#include <vector>

#define fast_io ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);
#define endl "\n"

using namespace std;

struct Node{
	int data;
	struct Node *prev, *next;
};

class LList{

	private:
		void split(struct Node *head, struct Node** fh, struct Node** sh){
			if((head) == NULL || (head) -> next == NULL){
				*fh = head;
				*sh = NULL;
				return; 
			}
			struct Node* slowptr = head;
			struct Node* fastptr = (head) -> next;
			while(fastptr != NULL){
				fastptr = fastptr -> next;
				if(fastptr != NULL){
					slowptr = slowptr -> next;
					fastptr = fastptr -> next;
				}
			}
			*fh = head;
			*sh = slowptr -> next;
			slowptr -> next -> prev = NULL;
			slowptr -> next = NULL;
			return;
		}
		struct Node* merge(struct Node* fh, struct Node* sh, struct Node **tail){
			struct Node *head = NULL;
			while(fh != NULL && sh != NULL){
				if(fh->data <= sh->data){
					if(head == NULL)
						head = fh;

					struct Node *t = fh -> next;
					fh -> next = sh;
					if(sh -> prev){
						sh -> prev ->next = fh;
					}
					sh -> prev = fh;
					fh = t;
				}
				else{
					if(head == NULL)
						head = sh;

					struct Node *t = sh -> next;
					sh -> next = fh;
					if(fh -> prev){
						fh -> prev ->next = sh;
					}
					fh -> prev = sh;
					sh = t;
				}
			}
			if(fh != NULL){
				while(fh -> next != NULL){
					fh = fh -> next;
				}
				*tail = fh;
			}
			else if(sh != NULL){
				while(sh -> next != NULL){
					sh = sh -> next;
				}
				*tail = sh;
			}
			
			return head;
		}

	public:
		struct Node *head, *tail;

		LList(){
			this -> head = NULL;
		}

		void insert_at_tail(int data){
			struct Node* newnode = new Node;
			newnode -> data = data;
			newnode -> prev = newnode -> next = NULL;
			this -> tail = newnode;
			if(this -> head == NULL){
				this -> head = newnode;
				return;
			}
			struct Node* temp = this -> head;
			while(temp->next != NULL){
				temp = temp -> next;
			}
			temp -> next = newnode;
			newnode -> prev = temp;
			return;
		}

		void insert_at_head(int data){
			struct Node* newnode = new Node;
			newnode -> data = data;
			newnode -> prev = NULL;
			newnode -> next = this -> head;
			if(this -> head)
				(this -> head) -> prev = newnode;
			else
				this -> tail = newnode;
			this -> head = newnode;
			return ;
		}

		bool search(int val){
			struct Node* temp = this -> head;
			while(temp != NULL){
				if(temp -> data == val){
					return true;
				}
				temp = temp -> next;
			}
			return false;
		}

		void remove(int val){
			struct Node* temp = this -> head;
			while(temp != NULL){
				if(temp -> data == val){
					break;
				}
				temp = temp -> next;
			}
			if(temp != NULL){
				if(temp -> prev == NULL){
					struct Node* delnode = temp;
					this -> head = temp -> next;
					if(this -> head)
						(this->head) -> prev = NULL;
					else
						this -> tail = NULL;
					free(delnode);
				}
				else{
					struct Node* delnode = temp;
					temp -> prev -> next = temp -> next;
					if(temp -> next)
						temp -> next -> prev = temp -> prev;
					else
						this -> tail = temp -> prev;
					free(delnode);
				}
			}
			this -> display();
			this -> displayReverse();
			return ;
		}

		void reverse(struct Node* head){
			if(head -> next == NULL){
				this -> head = head;
				return;
			}
			reverse(head -> next);
			head -> next -> next = head;
			head-> next = NULL;

		}

		void mergeSort(struct Node **head, struct Node **tail){
			if((*head) == NULL || (*head) -> next == NULL)
				return;
			struct Node *fHalf, *sHalf;
			split(*head, &fHalf, &sHalf);
			mergeSort(&fHalf, tail);
			mergeSort(&sHalf, tail);
			*head = merge(fHalf, sHalf, tail);
		}

		void display(){
			cout<<"null";
			struct Node* temp = this -> head;
			while(temp){
				cout<<" <-- "<<temp->data<<"--> ";
				temp = temp -> next;
			}
			cout<<"null"<<endl;
		}

		void displayReverse(){
			cout<<"null";
			struct Node* temp = this -> tail;
			while(temp){
				cout<<" <-- "<<temp->data<<" --> ";
				temp = temp -> prev;
			}
			cout<<"null"<<endl;
		}
			
};

int main(){

	fast_io;

	freopen("../input.txt","r",stdin);
	freopen("../output.txt","w",stdout);

	int n,i;
	cin>>n;
	int a[n];
	for(i=0; i<n; i++)
		cin>>a[i];

	LList slist;

	//Insertion 
	cout<<"insertion at tail"<<endl;
	for(int i=0; i<n; i++){
		slist.insert_at_tail(a[i]);
	}
	slist.display();
	slist.displayReverse();

	LList slist2;
	cout<<"insertion at head"<<endl;
	for(int i=0; i<n; i++){
		slist2.insert_at_head(a[i]);
	}
	slist2.display();
	slist2.displayReverse();

	//Search
	int searchVal;
	cin>>searchVal;
	bool isAvailable = slist.search(searchVal);
	if(isAvailable)
		cout<<searchVal<<" is found."<<endl;
	else
		cout<<searchVal<<" is not found."<<endl;

	//Delete
	int delVal;
	cin>>delVal;
	cout<<"After deletion of node "<<delVal<<endl;
	slist.remove(delVal);

	// Merge Sort
	slist.mergeSort(&slist.head, &slist.tail);
	cout<<"Sorted List"<<endl;
	slist.display();
	cout<<"Traversing Sorted List in descending order"<<endl;
	slist.displayReverse();
}

File: DataStructures/DoubleLinkedList/python/DoubleLinkedList.py
import sys

class Node:
	def __init__(self, data, prev = None, nexxt = None):
		self.data = data
		self.prev = prev
		self.next = nexxt

class DLList:
	def __init__(self):
		self.head = None
		self.tail = None

	def insert(self, data):
		newnode = Node(data)
		self.tail = newnode
		if self.head is None:
			self.head = newnode
		else:
			temp = self.head
			while temp.next is not None:
				temp = temp.next
			temp.next = newnode 
			newnode.prev = temp
		return

	def insert_at_head(self, data):
		newnode = Node(data)
		newnode.next = self.head 
		if self.head is not None:
			self.head.prev = newnode
		else:
			self.tail = newnode
		self.head = newnode
		return 

	def search(self, val):
		temp = self.head
		while(temp):
			if(temp.data == val):
				return True
			temp = temp.next
		return False

	def remove(self, val):
		temp = self.head
		while(temp):
			if(temp.data == val):
				break
			temp = temp.next
		if(temp):
			if(temp.prev is None):
				self.head = temp.next
				if self.head is not None:
					self.head.prev = None
				else:
					self.tail = None
			else:
				temp.prev.next = temp.next;
				if temp.next is not None:
					temp.next.prev = temp.prev;
				else:
					self.tail = temp.prev;
			temp = None
		self.display()

	def reverse(self, head):
		if head.next is None:
			self.head = head
			return
		self.reverse(head.next)
		head.next.next = head
		head.next = None

	def __split(self, head):    
		if head is None or head.next is None:
			return head, None
		slowptr = head
		fastptr = head.next
		while(fastptr):
			fastptr = fastptr.next
			if(fastptr):
				slowptr = slowptr.next
				fastptr = fastptr.next
		fHalf = head
		sHalf = slowptr.next
		slowptr.next.prev = None
		slowptr.next = None
		return fHalf, sHalf

	def __merge(self, a, b):
		temp = None
		while(a != None and b != None):
			if(a.data <= b.data):
				if(temp == None):
					temp = a;
				t = a.next
				a.next = b
				if(b.prev != None):
					b.prev.next = a
				
				b.prev = a
				a = t
			else:
				if(temp == None):
					temp = b
				t = b.next
				b.next = a
				if(a.prev != None):
					a.prev.next = b
				a.prev = b
				b = t
		if(a != None):
			while(a.next != None):
				a = a.next;
			self.tail = a;
		elif(b != None):
			while(b.next != None):
				b = b.next;
			self.tail = b;
		
		return temp;	

	def mergeSort(self, head):
		if head is None or head.next is None:
			return head
		fHalf, sHalf = self.__split(head)
		fHalf = self.mergeSort(fHalf)
		sHalf = self.mergeSort(sHalf)
		return self.__merge(fHalf, sHalf)

	def display(self):
		temp = self.head
		print("null",end = "")
		while(temp):
			print(" <-- ",temp.data,end = " --> ")
			temp = temp.next
		print("null")

	def displayReverse(self):
		temp = self.tail
		print("null",end = "")
		while(temp):
			print(" <-- ",temp.data,end = " --> ")
			temp = temp.prev
		print("null")

if __name__ == '__main__':

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = sys.stdin.readline();
	a = sys.stdin.readline().split(" ")[:-1] # [:-1] to eliminate reading of newline
	a = [int(i) for i in a]

	dlist = DLList()

	#Insertion
	print("Insertion at tail",end = "\n")
	for i in a:
		dlist.insert(i)
	dlist.display()
	dlist.displayReverse()

	dlist2 = DLList()
	print("Insertion at head",end = "\n")
	for i in a:
		dlist2.insert_at_head(i)
	dlist2.display()
	dlist2.displayReverse()

	#search
	sval = int(sys.stdin.readline())
	isAvailable = dlist.search(sval)
	if(isAvailable):
		print(sval," is found.")
	else:
		print(sval," is not found.")

	#delete
	dval = int(sys.stdin.readline())
	print("After deletion of node ", dval,end = "\n")
	dlist.remove(dval)

	# merge sort
	print("Sorted List", end = "\n")
	dlist.head = dlist.mergeSort(dlist.head)
	dlist.display()
	print("Traversing Sorted List in descending order", end = "\n")
	dlist.displayReverse()